src/Processors/QueryPlan/QueryPlan.cpp

```diff
@@ -502,8 +469,6 @@ void QueryPlan::optimize(const QueryPlanOptimizationSettings & optimization_sett
     QueryPlanOptimizations::optimizeTreeSecondPass(optimization_settings, *root, nodes);
     if (optimization_settings.build_sets)
         QueryPlanOptimizations::addStepsToBuildSets(*this, *root, nodes);
-
-    updateDataStreams(*root);
```

Comment:
Previously, this was used to update the query plan stream's properties (mainly about sorting). Now, streams are removed (and sorting property replaced with `applySorting` optimization), and this pass is not needed.

Meanwhile, this was hiding a bug in `optimizePrewhere` optimization: constant for the filter result could have been removed. In general, optimizations of QueryPlan should be local (only a few near steps are affected), and the headers for unchanged steps should be kept. Calling `updateInputHeaders` can partly restore this invariant, but hide potential bugs.

I will try to fix filtering by UInt8 column later, in a different PR.

Reply:
Decided to fix in the same PR